### PR TITLE
Reduce number of allocations during grain method call

### PR DESCRIPTION
**Pull request by dVakulen**
Currently on each grain method call string with debug context is being [allocated](https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/GrainReference.cs#L354). 
Sample run of 28k grain methods calls.
Before fix:
image
After:

### CODE REVIEW

**Comment by jthelin** [comment]
Nice job!

**Comment by ReubenBond** [comment]
Great work! :+1: If you feel like removing an additional, consider this one and tell me if you think it's sound.

```C#
protected async Task<T> InvokeMethodAsync<T>(int methodId, object[] arguments, InvokeMethodOptions options = InvokeMethodOptions.None, SiloAddress silo = null)
{
    object[] argsDeepCopy = null;
    if (arguments != null)
    {
        CheckForGrainArguments(arguments);
        argsDeepCopy = (object[])SerializationManager.DeepCopy(arguments);
    }
```

The above call to DeepCopy can be put into a loop, replacing the elements in the `arguments` array with their DeepCopy - that way we can reuse the array, which is allocated in generated `GrainReference` implementations (that method is `protected` - only to be used for that purpose).

Basically something like: `arguments[i] = DeepCopy(arguments[i])`.

What do you think?

**Comment by dVakulen** [comment]
@ReubenBond I'll remove this allocation, and below at https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/GrainReference.cs#L342 there is a call to https://github.com/dotnet/orleans/blob/master/src/Orleans/Async/TaskExtensions.cs#L226 which purpose besides calling ```if (task.Status == TaskStatus.Created) task.Start();``` I don't quite understand. If content of the method ```ConvertTaskViaTcs``` were inlined into ```InvokeMethodAsync``` that could help to avoid awaiting when task execution already finished, and pretty expensive call to ```ContinueWith``` also could be removed.

**Comment by gabikliot** [comment]
@ReubenBond , @dVakulen regarding the DeepCopy optimization:
We can't do that. The result would mean that we can break reference equality while serializing/deep copying. If the same object is referenced from 2 args, currently we will copy it once, preserving reference equality. With your suggestion, we will copy it twice.

**Comment by gabikliot** [comment]
@dVakulen , the purpose of `ConvertTaskViaTcs` is indeed rather cryptic and not well documented, but still very subtle. Shortly, it is meant to prevent Task execution inlining, which could cause deadlock.
You can get more details here: https://github.com/dotnet/orleans/pull/875  and https://github.com/dotnet/orleans/pull/853#issuecomment-143941910.

The summary, after all those discussions was, that we should wait till we move to .NET 4.6, which supports `TaskCreationOptions.RunContinuationsAsynchronously` and then this whole `ConvertTaskViaTcs` would be replaced with `TaskCreationOptions.RunContinuationsAsynchronously.

**Comment by dVakulen** [comment]
@gabikliot Thank you for the explanation.
